addon/hooks/pagination.ts

```diff
@@ -0,0 +1,283 @@
+     * Loads the next page of models if there are more to load and is not currently loading
+     * @returns {Promise<T[]> | null}
+     */
+    @action
+    loadMoreModels(): Promise<T[]> | null {
+        if (this.hasMore && !this.isLoadingModels) {
+            return this.loadModels();
+        }
+
+        return null;
+    }
+
+    /**
+     * Reloads the first page of models, alias for `loadModels(true)`
+     * @returns {Promise<T[]>}
+     */
+    @action
+    reloadModels() {
+        return this.loadModels(true);
+    }
+
+    /**
+     * Reloads the first page of models w/filters applied, alias for `loadModels(true)`
+     * @returns {Promise<T[]>}
+     */
+    @action
+    filterModels() {
+        return this.loadModels(true);
+    }
+
+    /**
+     * Clears all current model rows array
+     */
+    @action
+    clearModels() {
+        this.rows = A();
+    }
+
+    /**
+     * Deletes the model and removes it from the rows array
+     * @param {T} model
+     * @returns {Promise<void>}
+     */
+    @action
+    async removeModel(model: T) {
+        const result = await model.destroyRecord();
+        this.rows.removeObject(result);
+        return result;
+    }
+
+    /**
+     * Updates the current sorts, calls an onChangeSorting() handler if provided
+     * and reloads the first page of models
+     * @param {Sorting[]} newSorts
+     */
+    @action
+    async changeSorting(newSorts: Sorting[]) {
+        this.sorts = newSorts.map((col) =>
+            `${!col.isAscending ? '-' : ''}${col.sortPath ?? col.valuePath}`
+        );
+
+        //allow the parent context to store and/or modify updates to sorts
+        if(this.config.onChangeSorting) {
+            const processedSorts = await this.config.onChangeSorting(this.sorts, newSorts);
+            if(processedSorts) {
+                this.sorts = processedSorts;
+            }
+        }
+
+        return this.reloadModels();
+    }
+
+    /**
+     * Clears the current sorts and reloads the first page of models
+     */
+    @action
+    clearSorting() {
+        return this.changeSorting([]);
+    }
+
+    /**
+     * Clears all models from the rows array and resets the current state
+     * Sometimes useful in resetController() when the pagination may not
+     * be recreated/overwritten on every transition, and you want to clear
+     * it when leaving the page.
+     */
+    @action
+    reset() {
+        this.clearModels();
+        this.hasMore = true;
+        this.isLoading = false;
+    }
+}
+
+/**
+ * Creates and returns a new Pagination instance and binds its owner to be the same as
+ * that of its parent "context" (e.g. Controller, Component, etc).
+ * In most cases, this returned instance should be assigned to a @tracked property
+ * on its parent context, so that it can be accessed on the associated template
+ * @param {PaginationArgs} args
+ */
+const usePagination = <T extends DS.Model, M = ResponseMetadata>(args: PaginationArgs<T, M>) => {
```

Comment:
Not sure if we should go with the react `use` format here, or keep it like our Ember apps and just use `pagination`. I could go either way honestly

Reply:
Yeah, I thought about that a bit, and though we really don't have a pattern like this anywhere else in any of our ember apps, its a pretty unique/non-standard pattern (i.e. generating a class instance and manually binding it to the owner of a parent context), so I felt like coming up with something new (or hijacking react's convention) was maybe warranted here.

I don't love just calling it `pagination()` as the main backing class is called `Pagination`, so I wanted to differentiate it a bit more. `usePagination()` felt pretty good for what its doing, but i'm open to other ideas for a non-react convention if we don't want that confusion across tech stacks, maybe like `applyPagination()` or `bindPagination()`?

Reply:
I also think if we end up doing a "hook" type pattern like this for other things, it would be nice to have a common "verb prefix" convention on the main function for consistency.

Reply:
Yea that's fair. Im ok with keeping it how it is